web/thesauruses/javascript/ECMAScript 2009/classes.json

```diff
@@ -0,0 +1,108 @@
+{
+  "meta": {
+    "language": "javascript",
+    "language_version": "ECMAScript 2009",
+    "language_name": "JavaScript",
+    "structure": "classes"
+  },
+  "concepts": {
+    "normal_class": {
+      "code": "function ClassName() {\n // Constructor body\n}\nClassName.prototype.publicMethod = function() {\n // Method body\n}\nClassName.prototype.publicField = 0;",
+      "comment": "Classes are not present in ECMAScript 2009 and below versions but can be simulated using functions and prototypes.",
+      "name": "Normal class"
+    },
+    "abstract_class": {
+      "not-implemented": true,
+      "name": "Abstract class"
+    },
+    "interface": {
+      "not-implemented": true,
+      "name": "Interface"
+    },
+    "read_only_class": {
+      "not-implemented": true,
+      "name": "Read-only class"
+    },
+    "static_class": {
+      "not-implemented": true,
+      "name": "Static class"
+    },
+    "inner_class": {
+      "not-implemented": true,
+      "name": "Inner class"
+    },
+    "packages": {
+      "comment": "A package manager like NPM (https://www.npmjs.com/) or Deno (https://deno.land/) can be used to create and manage Javascript packages.",
+      "not-implemented": true,
+      "name": "Packages"
+    },
+    "class_with_generic_type": {
+      "not-implemented": true,
+      "name": "Class with a generic type"
+    },
+    "private_variables": {
+      "code": "var ClassName = (function() {\n var privateField = 0;\n // privateField is accessible in this scope only.\n function ClassName() {\n  // Constructor body\n }\n return ClassName;\n})();",
```

Comment:
[NB] What!? This is a thing? I had no clue you could do things like this in JS.

Reply:
It was a thing but today, Babel.js can convert from ES6+ to ES5. Furthermore, all browsers are now compatible with ES6+.

Reply:
I removed the private variables since there is no keyword "private" (which is not "private" in the last specification btw). The previous code is more like a trick than a solution.

Reply:
To know more about this syntax, a page on MDN about "Immediately Invoked Function Expression" is available there : https://developer.mozilla.org/en-US/docs/Glossary/IIFE